CIPs/0017.md

```diff
@@ -0,0 +1,165 @@
+                           // If batch verification should be used, newValidators and deletedValidators will be empty.
+  NewValidators     []blsCrypto.SerializedPublicKey // ValSetDiff(FirstEpoch, LastEpoch), added validators
+  DeletedValidators *big.Int // ValSetDiff(FirstEpoch, LastEpoch): bitmap of deleted validators from the `firstEpoch` valset.
+}
+```
+
+The light clients will need to store verification keys corresponding to the existing versions of the snark circuit, as elaborated in #upgradeability. 
+
+There are two proposed extensions for the `les` protocol:
+
+Option 1:
+This option is more resilient to incomplete proof propagation for full nodes, but requires an extra round trip of messages to achieve this. The idea is for the client to query the server to see what proofs are available, and then request the optimal proofs. If the client wants to minimize network traffic and assume proofs are well distributed, they can also bypass the `GetAvailableProofs` step and just request that the server sends the best proofs it has for the client's current epoch (using the `handleLogic` flag on `GetEpochHeadersOrProofs`).
+```
+GetAvailableProofs - Asks the server what proofs it has. Server responds with `[]proofMetadata` of all available proofs.
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, handleLogic boolean, requestedProofs []proofMetadata)
+    `highestEpoch` is the current highestEpoch the client knows of.
+    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
+                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
+    `requestedProofs`: Set of proofs' metadata requested by the client
+  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
+  Both responses are subject to message size limits limiting the number of proofs/headers.
+EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.
+```
+
+Option 2: This option benefits from assumptions on the pattern of proof generation (likely at least one for every ~120 epochs) and by assuming proofs are well distributed (all full nodes have all proofs). The idea here is to pass a bitmap to the `GetEpochHeadersOrProofs` msg that indicates if the client knows the validator set for the past `SIZE_OF_BITMAP` (~240?) epochs. This allows the full node to optimize the set of proofs sent back, while minimizing computation for the client. The main win here would be in a case like this: the client uses a proof to sync to epoch 120, then syncs further to epoch 121, so it knows the validator set of 120 and 121. Then the client goes down for > 120 epochs, and when it comes back the server can figure out to send a proof from 120-240 that we assume is generated, rather than sending 119 epoch headers to verify.
```

Comment:
I'm not sure that this approach actually relies on those assumptions. The light node gives the full node all the information it needs in order to construct a reasonably efficient batch of proofs/headers to sync the light node.

The light client could include in that request, the maximum number of proofs/headers it's willing to verify.

If the full node cannot meet that request, it returns an empty response. If it *can* meet the request, great!

We can have these values be configurable, such that they can be tightened up towards the optimal values of `(CEIL(epochs / 120), 0)`

Reply:
I'm personally in favor of there being one and only set of proofs and corresponding instance info per epoch that a server will serve. Basically, if clients go offline for a while and get sufficiently far behind, they should just start at the genesis epoch again and verify to the present that way. Maybe after 2 weeks or so downloading the non-SNARK epoch messages becomes more bandwidth/CPU intensive versus just re-bootstrapping with SNARKs. This also very much simplifies proof management these proof serving servers need to do, only storing the minimal number of proofs to get from the genesis to current epoch.

Reply:
Hmm I'm not sure I see why that's necessary. The light client can just ask to sync from the latest epoch N for which it has the validator set where e.g. `N % 120 == 0`.

Reply:
Yes, clarifications below.

Reply:
Agreed this approach doesn't rely on these assumptions, I was just noting that it benefits from them and produces a more optimal flow if they hold.

Maybe this idea of a max # proofs/headers can solve for the potential problem above of full nodes not having every proof? I'll draw out a potential network flow and think through some data structures for a couple solutions and see if I can decide on a solution